File: web/src/api/runtimeNodes.ts
import { call, http } from './client'
import type { RelayOut } from './types'
export interface Capability {
  installed: boolean; version: string; login: 'ready' | 'required' | 'unknown'
  health: string; detail: string; models: string[]
}
export interface RuntimeNode {
  id: string; name: string; hostname: string; username: string
  platform: string; architecture: string; environment: string; workspace_root: string
  version: string; online: boolean; is_active: boolean; draining: boolean
  heartbeat_at: string | null; service_status: string; team_id: string | null; team_name: string | null
  /** 节点并发上限与当前进行中的调用数；Daemon 未上报时为 null。 */
  max_concurrent: number | null; active_calls: number | null
  /** 节点 config.json 的 Git 主机白名单（只读）；旧版 Daemon 未上报时为 null。 */
  git_hosts: string[] | null
  capabilities: Record<string, Capability>; backends: RelayOut[]
}
export interface InstallLink {
  id: string; name: string; workspace_root: string; team_id: string | null
  state: 'ready' | 'used' | 'expired' | 'revoked'; expires_at: string; created_at: string
  node_id: string | null; command?: string; url?: string
}
export interface InstallInput {
  name: string; workspace_root: string; team_id: string | null
  /** ca_pem：私有 CA 证书（PEM，≤64 KB）；留空时不传该键。 */
  options: { control_proxy?: string; environment?: string; proxy: string; ca_pem?: string; claude_path: string; codex_path: string; git_hosts: string[]; max_concurrent: number; install_claude_probe: boolean }
}
const base = '/api/admin/runtime-nodes'
export const runtimeNodes = {
  list: () => call<RuntimeNode[]>(http.get(base)),
  /** team_id 传 null 表示改为公共池；不传表示不改团队。 */
  patch: (id: string, body: { name?: string; team_id?: string | null; is_active?: boolean; draining?: boolean }) => call<null>(http.patch(`${base}/${id}`, body)),
  remove: (id: string) => call<null>(http.delete(`${base}/${id}`)),
  createLink: (body: InstallInput) => call<InstallLink>(http.post(`${base}/install-links`, body)),
}

File: coreman/core/db/models/runtime_nodes.py
"""Self-enrolled runtime nodes and the cross-process reverse HTTP mailbox."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, Text, Uuid, text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from coreman.core.db.base import Base, TimestampMixin


class RuntimeNode(TimestampMixin, Base):
    __tablename__ = "runtime_nodes"
    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True)
    name: Mapped[str] = mapped_column(Text)
    token_hash: Mapped[str] = mapped_column(Text)
    team_id: Mapped[uuid.UUID | None] = mapped_column(Uuid, ForeignKey("teams.id"))
    visibility: Mapped[str] = mapped_column(Text, server_default="all")
    workspace_root: Mapped[str] = mapped_column(Text)
    root_change: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    root_edit_supported: Mapped[bool] = mapped_column(Boolean, server_default=text("false"))
    hostname: Mapped[str] = mapped_column(Text)
    username: Mapped[str] = mapped_column(Text)
    platform: Mapped[str] = mapped_column(Text)
    architecture: Mapped[str] = mapped_column(Text)
    environment: Mapped[str] = mapped_column(Text)
    version: Mapped[str] = mapped_column(Text)
    is_active: Mapped[bool] = mapped_column(Boolean, server_default=text("true"))
    draining: Mapped[bool] = mapped_column(Boolean, server_default=text("false"))
    heartbeat_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    capabilities: Mapped[dict[str, Any]] = mapped_column(JSONB, server_default="{}")
    service_status: Mapped[str] = mapped_column(Text, server_default="unknown")
    # 节点上报的协议版本；旧节点不上报时为空。
    protocol_version: Mapped[int | None] = mapped_column(Integer)
    # 节点在心跳里上报的并发上限与当前在执行数；旧节点不上报时为空。
    max_concurrent: Mapped[int | None] = mapped_column(Integer)
    active_calls: Mapped[int | None] = mapped_column(Integer)
    # 节点 config.json 的 git_hosts 白名单，由心跳上报、只读展示；旧节点不上报时为空。
    git_hosts: Mapped[list[str] | None] = mapped_column(JSONB)
    # CLI 继承到的出站代理 {source, url, pending}，由心跳上报、只读展示；旧节点不上报时为空。
    proxy: Mapped[dict[str, Any] | None] = mapped_column(JSONB)


class RuntimeInstallLink(TimestampMixin, Base):
    __tablename__ = "runtime_install_links"
    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    token_hash: Mapped[str] = mapped_column(Text)
    created_by: Mapped[uuid.UUID | None] = mapped_column(Uuid, ForeignKey("users.id"))
    name: Mapped[str] = mapped_column(Text)
    team_id: Mapped[uuid.UUID | None] = mapped_column(Uuid, ForeignKey("teams.id"))
    visibility: Mapped[str] = mapped_column(Text, server_default="all")
    workspace_root: Mapped[str] = mapped_column(Text)
    options: Mapped[dict[str, Any]] = mapped_column(JSONB, server_default="{}")
    expires_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    revoked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    node_id: Mapped[uuid.UUID | None] = mapped_column(Uuid, ForeignKey("runtime_nodes.id"))
    used_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class RuntimeCall(Base):
    __tablename__ = "runtime_calls"
    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    node_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("runtime_nodes.id", ondelete="CASCADE"), index=True
    )
    provider: Mapped[str] = mapped_column(Text)
    request_enc: Mapped[str] = mapped_column(Text)
    status: Mapped[str] = mapped_column(Text, server_default="queued")
    status_code: Mapped[int | None] = mapped_column(Integer)
    content_type: Mapped[str | None] = mapped_column(Text)
    next_seq: Mapped[int] = mapped_column(Integer, server_default="0")
    response_bytes: Mapped[int] = mapped_column(Integer, server_default="0")
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=text("now()")
    )
    deadline: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True)
    consumer_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    error: Mapped[str | None] = mapped_column(Text)


class RuntimeChunk(Base):
    __tablename__ = "runtime_chunks"
    call_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("runtime_calls.id", ondelete="CASCADE"), primary_key=True
    )
    seq: Mapped[int] = mapped_column(Integer, primary_key=True)
    data_enc: Mapped[str] = mapped_column(Text)
